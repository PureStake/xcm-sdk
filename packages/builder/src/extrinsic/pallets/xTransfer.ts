/* eslint-disable sort-keys */
import { ExtrinsicConfig } from '../ExtrinsicConfig';
import { ExtrinsicConfigBuilder } from '../ExtrinsicConfigBuilder.interfaces';

const pallet = 'xTransfer';

export function xTransfer() {
  return {
    transfer: () => {
      const method = 'transfer';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: ({ address, amount, destination }) =>
            new ExtrinsicConfig({
              pallet,
              method,
              args: [
                {
                  id: {
                    Concrete: {
                      parents: 0,
                      interior: 'Here',
                    },
                  },
                  fun: {
                    Fungible: amount,
                  },
                },
                {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      {
                        AccountKey20: {
                          key: address,
                        },
                      },
                    ],
                  },
                },
                {
                  refTime: 5_000_000_000,
                  proofSize: 0,
                },
              ],
            }),
        }),
        X2: (): ExtrinsicConfigBuilder => ({
          build: ({ address, amount, destination, asset }) =>
            new ExtrinsicConfig({
              pallet,
              method,
              args: [
                {
                  id: {
                    Concrete: {
                      parents: 1,
                      interior: {
                        X2: [
                          {
                            Parachain: destination.parachainId,
                          },
                          {
                            PalletInstance: asset,
                          },
                        ],
                      },
                    },
                  },
                  fun: {
                    Fungible: amount,
                  },
                },
                {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      {
                        AccountKey20: {
                          key: address,
                        },
                      },
                    ],
                  },
                },
                {
                  refTime: 5_000_000_000,
                  proofSize: 0,
                },
              ],
            }),
        }),
      };
    },
  };
}
