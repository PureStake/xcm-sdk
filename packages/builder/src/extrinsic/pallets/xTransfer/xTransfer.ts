/* eslint-disable sort-keys */
import { ExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import { ExtrinsicConfig } from '../../ExtrinsicConfig';

const pallet = 'xTransfer';

export function xTransfer() {
  return {
    transfer: () => {
      const method = 'transfer';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: ({ address, amount, destination }) =>
            new ExtrinsicConfig({
              pallet,
              func: method,
              getArgs: () => [
                {
                  id: {
                    Concrete: {
                      parents: 0,
                      interior: 'Here',
                    },
                  },
                  fun: {
                    Fungible: amount,
                  },
                },
                {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      {
                        AccountKey20: {
                          key: address,
                        },
                      },
                    ],
                  },
                },
                {
                  refTime: 5_000_000_000,
                  proofSize: 0,
                },
              ],
            }),
        }),
        X2: (): ExtrinsicConfigBuilder => ({
          build: ({ address, amount, destination, asset }) =>
            new ExtrinsicConfig({
              pallet,
              func: method,
              getArgs: () => [
                {
                  id: {
                    Concrete: {
                      parents: 1,
                      interior: {
                        X2: [
                          {
                            Parachain: destination.parachainId,
                          },
                          {
                            PalletInstance: asset,
                          },
                        ],
                      },
                    },
                  },
                  fun: {
                    Fungible: amount,
                  },
                },
                {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      {
                        AccountKey20: {
                          key: address,
                        },
                      },
                    ],
                  },
                },
                {
                  refTime: 5_000_000_000,
                  proofSize: 0,
                },
              ],
            }),
        }),
      };
    },
  };
}
