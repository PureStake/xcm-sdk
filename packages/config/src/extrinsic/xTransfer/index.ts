export * from './xTransfer';
export * from './xTransfer.constants';
export * from './xTransfer.interfaces';
