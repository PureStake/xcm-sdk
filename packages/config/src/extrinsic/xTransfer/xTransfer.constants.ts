export enum XTransferExtrinsic {
  Transfer = 'transfer',
}

export enum XTransferExtrinsicSuccessEvent {
  Deposited = 'Deposited',
  Withdrawn = 'Withdrawn',
}
