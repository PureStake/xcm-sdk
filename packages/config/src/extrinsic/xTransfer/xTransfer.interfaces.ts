import { Parents, XcmExtrinsicGetParams } from '../common.interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  XTransferExtrinsic,
  XTransferExtrinsicSuccessEvent,
} from './xTransfer.constants';

export interface XTransferPallet {
  pallet: ExtrinsicPallet.XTransfer;
  extrinsic: XTransferExtrinsic;
  successEvent: XTransferExtrinsicSuccessEvent;
  getParams: (params: XcmExtrinsicGetParams) => XTransferPalletParams;
}

export type XTransferPalletParams = [
  /**
   * asset
   */
  {
    id: {
      Concrete: XTransferConcreteParam;
    };
    fun: {
      Fungible: bigint;
    };
  },
  /**
   * destination
   */
  {
    parents: 1;
    interior: {
      X2: [
        {
          Parachain: number;
        },
        {
          AccountKey20: {
            /**
             * account
             */
            key: string;
          };
        },
      ];
    };
  },
  /**
   * weight
   */
  {
    refTime: 5_000_000_000;
    proofSize: 0;
  },
];

export interface XTransferConcreteParam {
  parents: Parents;
  interior: XTransferInteriorParam;
}

export type XTransferInteriorParam =
  | 'Here'
  | {
      X2: [
        {
          Parachain: number;
        },
        {
          PalletInstance: number;
        },
      ];
    };
