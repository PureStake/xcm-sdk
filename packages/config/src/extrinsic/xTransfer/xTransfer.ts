import { ChainConfig, MoonChainConfig } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  XTransferExtrinsic,
  XTransferExtrinsicSuccessEvent,
} from './xTransfer.constants';
import {
  XTransferConcreteParam,
  XTransferPallet,
} from './xTransfer.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function xTransfer(chain: MoonChainConfig) {
  return {
    transfer: () => transfer(chain),
  };
}

function transfer(chain: MoonChainConfig) {
  return {
    successEvent: (event: XTransferExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          XTransferExtrinsic.Transfer,
          event,
          chain,
          origin,
        );

        return {
          here: () =>
            createExtrinsic(() => ({
              parents: 0,
              interior: 'Here',
            })),
          X2: (palletInstance: number) =>
            createExtrinsic(() => ({
              parents: 1,
              interior: {
                X2: [
                  {
                    Parachain: chain.parachainId,
                  },
                  {
                    PalletInstance: palletInstance,
                  },
                ],
              },
            })),
        };
      },
    }),
  };
}

function getCreateExtrinsic(
  extrinsic: XTransferExtrinsic,
  event: XTransferExtrinsicSuccessEvent,
  chain: MoonChainConfig,
  origin: ChainConfig,
) {
  return (getConcrete: () => XTransferConcreteParam): XTransferPallet => ({
    pallet: ExtrinsicPallet.XTransfer,
    extrinsic,
    successEvent: event,
    getParams: (account, amount) => [
      {
        id: {
          Concrete: getConcrete(),
        },
        fun: {
          Fungible: amount,
        },
      },
      {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: chain.parachainId,
            },
            {
              AccountKey20: {
                network: 'Any',
                key: account,
              },
            },
          ],
        },
      },
      origin.weight,
    ],
  });
}
