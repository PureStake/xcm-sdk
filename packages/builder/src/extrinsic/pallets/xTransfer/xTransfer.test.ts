import { buildParamsMock } from '../../../../fixtures';
import { xTransfer } from './xTransfer';

describe('xTransfer', () => {
  describe('transfer', () => {
    describe('here', () => {
      const extrinsic = xTransfer().transfer().here().build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x2', () => {
      const extrinsic = xTransfer().transfer().X2().build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });
});
