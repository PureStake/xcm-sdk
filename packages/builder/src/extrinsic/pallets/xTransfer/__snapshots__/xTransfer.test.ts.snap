// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xTransfer transfer here should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTransfer",
  "type": "Substrate",
}
`;

exports[`xTransfer transfer here should get correct arguments 1`] = `
[
  {
    "fun": {
      "Fungible": 99000000000n,
    },
    "id": {
      "Concrete": {
        "interior": "Here",
        "parents": 0,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;

exports[`xTransfer transfer x2 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTransfer",
  "type": "Substrate",
}
`;

exports[`xTransfer transfer x2 should get correct arguments 1`] = `
[
  {
    "fun": {
      "Fungible": 99000000000n,
    },
    "id": {
      "Concrete": {
        "interior": {
          "X2": [
            {
              "Parachain": 1000,
            },
            {
              "PalletInstance": "USDT",
            },
          ],
        },
        "parents": 1,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;
